### PR TITLE
JDK-8300808: Accelerate Base64 on x86 for AVX2

### PR DESCRIPTION
**Pull request by asgibbons**
Added code for Base64 acceleration (encode and decode) which will accelerate ~4x for AVX2 platforms.

Encode performance:
**Old:**
```
Benchmark                      (maxNumBytes)   Mode  Cnt     Score   Error   Units
Base64Encode.testBase64Encode           1024  thrpt    3  4309.439 ± 2.632  ops/ms
```

**New:**
```
Benchmark                      (maxNumBytes)   Mode  Cnt      Score     Error   Units
Base64Encode.testBase64Encode           1024  thrpt    3  24211.397 ± 102.026  ops/ms

```
Decode performance:
**Old:**
```
Benchmark                      (errorIndex)  (lineSize)  (maxNumBytes)   Mode  Cnt     Score    Error   Units
Base64Decode.testBase64Decode           144           4           1024  thrpt    3  3961.768 ± 93.409  ops/ms

**New:**
Benchmark                      (errorIndex)  (lineSize)  (maxNumBytes)   Mode  Cnt      Score    Error   Units
Base64Decode.testBase64Decode           144           4           1024  thrpt    3  14738.051 ± 24.383  ops/ms
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300808](https://bugs.openjdk.org/browse/JDK-8300808): Accelerate Base64 on x86 for AVX2

### Reviewers
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**) ⚠️ Review applies to [424b40d7](https://git.openjdk.org/jdk/pull/12126/files/424b40d75e416a75ec8f6d2abc1ae7db3182a71f)
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to [424b40d7](https://git.openjdk.org/jdk/pull/12126/files/424b40d75e416a75ec8f6d2abc1ae7db3182a71f)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12126/head:pull/12126` \
`$ git checkout pull/12126`

Update a local copy of the PR: \
`$ git checkout pull/12126` \
`$ git pull https://git.openjdk.org/jdk pull/12126/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12126`

View PR using the GUI difftool: \
`$ git pr show -t 12126`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12126.diff">https://git.openjdk.org/jdk/pull/12126.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (cl4es) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2661]: Sorry in advance since I'm probably reading this wrong: the data that `r13` is pointing to appears to be a repeated byte pattern (`0x2f2f2f...`), does this mean this `vpbroadcastb` and the `vpbroadcastq` above end up filling up their respective registers with the exact same bits? If so, and since neither of them is mutated in the code below, then perhaps this can be simplified a bit.

Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2661]: You're reading it correctly - this is redundant and could be handled differently, as the same value is being loaded into ymm4 and ymm10.  I don't think there will be any significant performance gain either way.  This was done in this manner to allow easier transition to URL acceleration when it is implemented, as URLs require handling '-' and '_' instead of '+' and '/' ('/' = 0x2f).

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (cl4es) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2673]: This got me curious (sorry!) and maybe there's something going on here that I'm not getting... But why have you split the loop apart and jump into the middle of it? It'd seem not doing this would be more straightforward, with no difference semantically and one less jump? (align32 should just translate to a narrow nop instruction, if anything)

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (cl4es) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2159]: A casual reader might wonder why there's 3 other avx2-tables and then this, so for readability it'd be nice to add "decode" to the name of this new table.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (cl4es) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2643]: Nevermind, you're not using r12 and GHA Windows testing is green now.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (cl4es) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2202]: Shouldn't this be `decode_lut_tables`? As it's used for URL and non-URL decoding alike.

Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2202]: These tables are used for both URL and non-URL based on the parameter, and they are only two of the three lut tables used (the other is in `base64_AVX2_decode_tables_addr` ).  Both names are essentially incorrect.  Does the name really matter that much?  It's the same as `base64_AVX2_decode_tables_addr` with the addition of URL tables.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2661]: Done.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (cl4es) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2202]: Names are important, but always hard to get right. At the very least they need to be correct. Maybe call it something like `..parameterized_decode_tables..` and the other `..shared_decode_tables..`?

Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2202]: I prefer leaving them the way they are.  I don't think the names, along with the associated comments within the tables, causes any undue confusion as to their function.  However I will implement the name change if that's all it takes to procure a review approval.  Please provide the specific names you'd like me to use and I'll change them.  Or just approve as-is :-).

Comment (cl4es) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2202]: I meant no disrespect here. By your own words `base64_AVX2_decode_URL_tables_addr` is "essentially incorrect", so I suggested some alternatives that I thought would make the code slightly more approachable. Regardless of whether you fix this detail I am not ready to approve this PR since I would need time to digest the latest changes. I'll ask a more senior engineer to review and give final approval (these changes need 2 reviewers approval anyhow).

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2693]: Subtraction effects EFLAGs we can save one redundant compare per iteration on [#L2697](https://github.com/openjdk/jdk/pull/12126/files#diff-b938ab8a7bd9f57eb02271e2dd24a305bca30f06e9f8b028e18a139c4908ec92R2697)
by doing a prior subtraction by 0x2c (44) in pre-loop and increment by same amount post loop.

Same goes out for encode main loop also.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2720]: For isURL, it looks to me that the vptest will fail for URL valid input 0x5F ("_"):
  upper_nibble =  0x5; 
  lower_nibble = 0xF;
  lut_lo_URL = 0x1B; (corresponding to 0xF)
  lut_hi = 0x8; (corresponding to 0x5)
  lut_lo_URL & lut_hi = 0x8; (not zero, taken as not allowable and so exit from loop)
Could you please verify on your end and fix this?
My understanding is that this is happening because 5 and 7 upper nibble get the same encoding 0x8.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2227]: The lut_roll URL doesn't seem to be correct:
0x5F (URL base64 ASCII for "/") would need an offset of -20H i.e. 0xEC.
However the others with upper nibble as 5 need an offset of -65H i.e. 0xBF.

It looks to me that the adjustment for 5F should be -4 instead of -1 at line 2722.

**Comment by asgibbons** [comment]
@sviswa7 Can you please share the test(s) you used to determine the stated failures?  I've run all of the tier1-3 tests in the suite with no failures.  Plus, if what you say is true, then the same logic would apply for the non-URL case as well as it applies to 0x2F ('/') in the same manner as 0x5F ('_').

If your test(s) do indeed show the failure, we should probably add them to the test suite.  Thanks.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2227]: @asgibbons Currently the base64 string with "_" is not getting the vectorized implementation. as per my previous comment. It should show up in the jmh test. Once you fix that you will come across this bug. The logic is wrong here for 5F case.

**Comment by jatin-bhateja** [comment]
> @sviswa7 Can you please share the test(s) you used to determine the stated failures? I've run all of the tier1-3 tests in the suite with no failures. Plus, if what you say is true, then the same logic would apply for the non-URL case as well as it applies to 0x2F ('/') in the same manner as 0x5F ('_').
> 
> If your test(s) do indeed show the failure, we should probably add them to the test suite. Thanks.

Stub processes the input till last legal byte and falls over to java implementation for error handling, thus URL decoder will not find any anomaly on encountering '_' or "-" in input sequence, the behavior will still be functionally correct. Problem will only show up in JMH micro as a perf degradation, our micro handles MIME and Base64 decoder can you kindly extend it for URL decoder with '-' and '_'  in input sequence.

**Comment by asgibbons** [comment]
@cl4es @sviswa7 @jatin-bhateja I believe I've addressed all of your concerns - thank you all for the reviews.  Please review when you can.  Thanks!

**Review by jatin-bhateja — APPROVED** [review]
Thanks @asgibbons.

**Review by cl4es — APPROVED** [review]

**Comment by asgibbons** [comment]
@cl4es Can you please initiate the in-depth testing required for this fix?  Thanks.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2399]: The avx2 code needs the length to be atleast 0x2c (44) bytes. We could directly go to non-avx code instead of L_bruteForce here. We will save one subtract/branch.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2658]: This could be Assembler::less instead of Assembler::lessEqual.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2699]: This could be Assembler::less instead of Assembler::lessEqual.

**Comment by cl4es** [comment]
> @cl4es Can you please initiate the in-depth testing required for this fix? Thanks.

I've checked out the sources and fired off a sanity run of the first couple of tiers. Holding off on testing higher tiers until @sviswa7's concerns has been resolved.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2399]: Done.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (cl4es) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2699]: I think with `lessEqual` we'll jump to `L_tailProc` for the final 32-byte chunk in inputs that are divisible by 32 (starting from 64), no? Using `less` avoids that, while not affecting performance of any other inputs.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2661]: The thought is that when the length is equal to 44 bytes, we could do the vector loop once before tail processing. The rest of the logic seems to allow that. 44 bytes of base64 -> 33 bytes decoded. So a 32 byte write in vector loop would still be ok and we wont be writing beyond.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2699]: As Claes mentioned, this would allow us to do one more iteration of vector loop.

**Review by sviswa7 — APPROVED** [review]
Looks good to me.

**Comment by cl4es** [comment]
I've started tier1-5 testing internally. Will let you know if we find any issues.

**Comment by sviswa7** [comment]
Thanks a lot @cl4es for doing the tests for this PR.

**Comment by cl4es** [comment]
No problem! Testing looks good.

/sponsor

**Pull request closed without merging**
